README.rst
===================
sc.galleria.support
===================

.. contents:: Table of Contents
   :depth: 2


Overview
--------

**sc.galleria.support** is a Plone package (add-on) providing simple gallery integration for Plone.

Using in a Plone Site
----------------------

Step 1: Activate it
^^^^^^^^^^^^^^^^^^^^

Go to the 'Site Setup' page in the Plone interface and click on the
'Add/Remove Products' link.

Choose the product **sc.galleria.support** (check checkbox at its left side)
and click the 'Activate' button.

Step 2: Setting preferences
^^^^^^^^^^^^^^^^^^^^^^^^^^^

In 'Site setup' below on 'Add-on Configuration' click on 'Galleria' option. In
this section you can see several options wich are categorized in different
types. Each type is visualized in a tab. You have options for 'Default',
'Flickr Plugin', 'Picasa Plugin' and 'History Plugin'. To know how to setup this
preferences in a way to go better to your needs see `Functionality`_ section.

Step 3: Seeing it in action
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Galleria will be rendered through a set of the display action in a container
content type. For example, if you have a folder named "My folder" and you are
standing there, click on "Display" (you need permissions for the edit bar) and then in "Galleria".
Of course, this will only have an effect if you have images inside that folder. If you do, you will
see the images render with the Galleria plugin.

For the case that you want to fetch pictures located in flickr or picasa see
`Flickr & Picasa Plugins`_. To activate Galleria in the Link just follow
the same procedure with the display action.

Functionality
--------------

The next explanations about different options must be applied in the control panel of Galleria
add-on. See `Step 2: Setting preferences`_.

Default options
^^^^^^^^^^^^^^^

In this section you have all the basic options to set how the layout of your
galleria will look like.


**Auto play:**
    This will start playing the slideshow with 5 seconds interval.

**Show informations:**
    Displays the caption.

**Gallery width:**
    You can use this option to set a gallery width manually. Default is 500 px.

**Gallery height:**
    You can use this option to set a gallery height manually. Default is 500 px.

**Image css position:**
    Positions the main image inside the stage container. Works like the CSS background-position 
    property, f.ex ‘top right’ or ‘20% 100%’. You can use keywords, percents or pixels. The first
    value is the horizontal position and the second is the vertical.

     - Read more about positioning at http://www.w3.org/TR/REC-CSS1/#background-position

**Enable lightbox:**
    This option acts as a helper for attaching a lightbox when the user clicks on an image. If you
    have a link defined for the image, the link will take precedence.

**Show counting:**
    Displays the counter.

**Transitions:**
    The transition that is used when displaying the images. There are different transitions in Galleria.

    Flavors are:
         - *fade* - crossfade betweens images.
         - *flash* - fades into background color between images.
         - *pulse* - quickly removes the image into background color, then fades the next image.
         - *slide* - slides the images depending on image position.
         - *fadeslide* - fade between images and slide slightly at the same time.

**Transition Speed:**
    The milliseconds used in the animation when applying the transition. The higher number, the slower transition.

**Wait:**
    Sets how long Galleria should wait when trying to extract measurements, before throwing an error. Default is 5000

**Show image navigation:**
    Displays the image navigation (next/prev arrows).

**Swipe:**
    Enables a swipe movement for flicking through images on touch devices.

**Selector jQuery:**
    Eg. '#content-core' or '#content' or '.galleria'. Do not change if you do not know what I mean.

**Show Thumbnails:**
    Sets the creation of thumbnails.

**Enable debug mode:**
    This option is for turning debug on/off. By default, Galleria displays errors by printing them out in the
    gallery container and sometimes throw exceptions. For deployment you can turn debug off to generate a more generic
    error message if a fatal error is raised.


YouTube, Vimeo & DailyMotion supported
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To use this supports you have to create a 'Link' content type and in the 'url'
field set the movie url.

 - **YouTube:** Sets options for the YouTube player.

 - **Vimeo:** Sets options for the Vimeo player.

 - **DailyMotion** Adds player options for the Daliymotion video player.

Flickr & Picasa Plugins
^^^^^^^^^^^^^^^^^^^^^^^

To use this plugins you have to create a 'Link' content type and in the 'url'
field set the galleria url that looks something like:

 - **Flickr:** 'http://www.flickr.com/photos/user_id/sets/galleria_id/'

 - **Picasa:** 'https://picasaweb.google.com/user_id/galleria_id'

**Enable plugin:** activate this function.

**Maximum number of photos:** you can set the maximum of photos to show.

**Show description:** Fetch the description. The plugin fetches the title per
default.

History Plugin
^^^^^^^^^^^^^^

The Galleria History plugin is a simple extension to create Galleria add hash
tags for permalinks and back button functionality enabled. This is useful on
fullscreen views and other use cases. The plugin simply adds a #/[id] hash to
the URL and then applies the necessary code for all browsers to enable the back
button. It also makes permalinks possible by simply bookmarking f.ex
http://mygalleria.com/#/4 and the user will be shown the 5th image in the
gallery (index starts at 0).

Browser support includes Firefox 2+, IE6+, Ipad, Opera and Chrome.

You enable it with the **Enable history plugin** option.


docs/HISTORY.txt
Changelog
----------

1.0.2 (unreleased)
^^^^^^^^^^^^^^^^^^

- Nothing changed yet.


1.0.1 (2017-01-11)
^^^^^^^^^^^^^^^^^^

- Deprecate package.
  [hvelarde]

- Fix package uninstall.
  [rodfersou, hvelarde]


1.0 (2013-01-27)
^^^^^^^^^^^^^^^^^^^

* Refactor travis-ci integration and tests.
  [cleberjsantos]

* Add upgrade steps.
  [cleberjsantos]

* Refactor to integration of collective.js.galleria.
  [cleberjsantos]

* Fix FaceBook plugin.
  [cleberjsantos]


1.0rc1 (2012-09-07)
^^^^^^^^^^^^^^^^^^^

* Added toggle full screen.
  [cleberjsantos]

* Added FaceBook Plugins.
  [cleberjsantos]

* Galleria upgrade version to 1.2.8 and templates fixes.
  [cleberjsantos]

* Refactoring tests.
  [cleberjsantos]

* Corrected thumb and image generation.
  [cleberjsantos]

* Added Galleria template to new collection.
  [cleberjsantos]

* Upgrade Picasa, Flickr and History Plugins.
  [cleberjsantos]


1.0b4 (2012-06-21)
^^^^^^^^^^^^^^^^^^

* Added image crop and responsive options.
  [cleberjsantos]

* Refactored and optimized code.
  [cleberjsantos]

* Refactoring browser test cases.
  [aleGpereira]

* Fixed Title and Description data, modification of the method call.
  [cleberjsantos]

* Resolves (issue `#25`_).
  [cleberjsantos]

* Added template with ZPT macro.
  [cleberjsantos]

* Updated documentation.
  [cleberjsantos]

* Updated templates.
  [cleberjsantos]

* Some tests were fixed and some others, refactored.
  [hvelarde]

* Fixed package distribution and documentation.
  [hvelarde]


1.0b3 (2012-05-28)
^^^^^^^^^^^^^^^^^^

* Fixed NameError: name 'test' is not defined (issue `#17`_).
  [cleberjsantos]

* PEP 8.
  [hvelarde]

* Portuguese translation updated.
  [lepri]


1.0b2 (2012-05-19)
^^^^^^^^^^^^^^^^^^

* Fixed Picasa and Flickr plugins (issue `#11`_).
  [cleberjsantos]

* Added validation for Vimeo and DailyMotion (issue `#12`_).
  [cleberjsantos]

* Patch for bug in Picasa plugin (issue `#13`_).
  [cleberjsantos]

* Added tests.
  [aleGpereira]

* Updated i18n and Brazilian Portuguese translation.
  [hvelarde]


1.0b1 (2012-05-18)
^^^^^^^^^^^^^^^^^^

*  Initial release.

.. _`#11`: https://github.com/simplesconsultoria/sc.galleria.support/issues/11
.. _`#12`: https://github.com/simplesconsultoria/sc.galleria.support/issues/12
.. _`#13`: https://github.com/simplesconsultoria/sc.galleria.support/issues/13
.. _`#17`: https://github.com/simplesconsultoria/sc.galleria.support/issues/17
.. _`#25`: https://github.com/simplesconsultoria/sc.galleria.support/issues/25


src/sc/galleria/support/interfaces.py
from zope.interface import Interface
from zope import schema

from z3c.form import field
from z3c.form import group

from zope.schema.vocabulary import SimpleVocabulary, SimpleTerm

import logging
logger = logging.getLogger('sc.galleria.support')

# dependencies
# Thanks: collective.gallery
try:
    #plone4
    from plone.app.folder.folder import IATUnifiedFolder as IFolder
    from Products.ATContentTypes.interfaces.link import IATLink as ILink
    from Products.ATContentTypes.interfaces.topic import IATTopic as ITopic
    from Products.ATContentTypes.interfaces.image import IATImage as IImage
except ImportError, e:
    logger.info('switch to plone3 %s' % e)
    #plone3
    from Products.ATContentTypes.interface import IATFolder as IFolder
    from Products.ATContentTypes.interface import IATLink as ILink
    from Products.ATContentTypes.interface import IATTopic as ITopic
    from Products.ATContentTypes.interface import IATImage as IImage

from sc.galleria.support import MessageFactory as _

transitionsvoc = SimpleVocabulary(
    [SimpleTerm(value='fade', title=_(u'Fade')),
     SimpleTerm(value='flash', title=_(u'Flash')),
     SimpleTerm(value='pulse', title=_(u'Pulse')),
     SimpleTerm(value='slide', title=_(u'Slide')),
     SimpleTerm(value='fadeslide', title=_(u'FadeSlide')), ]
)

thumbnailsvoc = SimpleVocabulary(
    [SimpleTerm(value='show', title=_(u"Show thumbnails")),
     SimpleTerm(value='empty', title=_(u"Don't show thumbnails")), ]
)


class IGalleriaLayer(Interface):
    """
    Marker Default browser layer this product.
    """


class IGalleria(Interface):
    """
    """

    def __init__(self, context, request, *args, **kwargs):
        """ """

    def galleriajs(self):
        """ """

    def getThumbnails(self):
        """ """

    def get_theme(self):
        """ """

    def portal_url(self):
        """ """

    def galleria_flickrid(self):
        """ """

    def galleria_picasauserandid(self):
        """ """


class IGeneralSettings(Interface):
    """Some general settings.
       These fields will appear on the 'Default' tab.
       Option informations: http://galleria.io/docs/1.2/options/
    """

    autoplay = schema.Bool(title=_(u"Auto Play."),
                           description=_(u"Sets Galleria to play slidehow when initialized."),
                           default=True,
                           required=True,)

    gallery_wait = schema.Int(title=_(u"Gallery Wait"),
                              description=_(u"Sets how long Galleria should wait when trying to extract measurements."),
                              default=5000,
                              required=True,)

    showInf = schema.Bool(title=_(u"Show informations"),
                          description=_(u"Toggles the caption."),
                          default=True,
                          required=True,)

    gallery_width = schema.Int(title=_(u"Gallery width"),
                               description=_(u"Manually set a gallery width."),
                               default=500,
                               required=True,)

    gallery_height = schema.Int(title=_(u"Gallery height"),
                                description=_(u"Manually set a gallery height."),
                                default=500,
                                required=True,)

    imagePosition = schema.TextLine(title=_(u"Image css position"),
                                    description=_(u"Eg. 'top right' or '20% 100%'"),
                                    default=_(u'center'),
                                    required=True,)

    lightbox = schema.Bool(title=_(u"Enable lightbox"),
                           default=False,
                           required=True,)

    showCounting = schema.Bool(title=_(u"Show counting"),
                               description=_(u"Toggles the counter."),
                               default=True,
                               required=True,)

    transitions = schema.Choice(title=_(u"Transitions"),
                                description=_(u"Defines what transition to use."),
                                default=_(u'fade'),
                                vocabulary=transitionsvoc,
                                required=True,)

    transitionSpeed = schema.Int(title=_(u"Transition Speed"),
                                 description=_(u"Defines the speed of the transition."),
                                 default=400,
                                 required=True,)

    showimagenav = schema.Bool(title=_(u"show image navigation"),
                               description=_(u"toggles the image navigation arrows."),
                               default=True,
                               required=True,)

    swipe = schema.Bool(title=_(u"swipe"),
                        description=_(u"Enables a swipe movement for flicking through images on touch devices."),
                        default=True,
                        required=True,)

    selector = schema.TextLine(title=_(u"Selector jQuery"),
                               description=_(u"Eg. '#content-core' or '#content' or '.galleria'. Do not change if you do not know what I mean."),
                               default=u"#content-galleria",
                               required=True,)

    thumbnails = schema.Choice(title=_(u"Show Thumbnails"),
                               description=_(u"Sets the creation of thumbnails"),
                               default=_(u'show'),
                               vocabulary=thumbnailsvoc,
                               required=True,)

    imagecrop = schema.Bool(title=_(u"Enable image crop"),
                            description=_(u"Defines how the main image will be cropped inside it is container."),
                            default=True,
                            required=True,)

    responsive = schema.Bool(title=_(u"Sets Gallery in responsive mode"),
                             description=_(u"Means that it will resize the entire container in dynamic proportions added in your CSS."),
                             default=True,
                             required=True,)

    debug = schema.Bool(title=_(u"Enable debug mode"),
                        description=_(u"Set this to false to prevent debug messages."),
                        default=False,
                        required=True,)


class IFaceBookPlugin(Interface):
    """ Enable/Disable FaceBook plugin
    """

    facebook = schema.Bool(title=_(u"Enable facebook plugin"),
                           description=_(u""),
                           default=False,)

    facebook_max = schema.Int(title=_(u"Maximum number of photos."),
                              description=_(u"Maximum number of photos to return (maximum value 100)."),
                              default=20,
                              required=True,)

    facebook_desc = schema.Bool(title=_(u"Show Description"),
                                description=_(u"The plugin fetches the title per default. If you also wish to fetch the description, set this option to true."),
                                default=False)


class IFlickrPlugin(Interface):
    """ Enable/Disable Flickr plugin
        http://galleria.io/docs/1.2/plugins/flickr/
    """

    flickr = schema.Bool(title=_(u"Enable flickr plugin"),
                         description=_(u""),
                         default=False,)

    flickr_max = schema.Int(title=_(u"Maximum number of photos."),
                            description=_(u"Maximum number of photos to return (maximum value 100)."),
                            default=20,
                            required=True,)

    flickr_desc = schema.Bool(title=_(u"Show Description"),
                              description=_(u"The plugin fetches the title per default. If you also wish to fetch the description, set this option to true."),
                              default=False)


class IPicasaPlugin(Interface):
    """ Enable/Disable Picasa plugin
        http://galleria.io/docs/1.2/plugins/picasa/
    """

    picasa = schema.Bool(title=_(u"Enable picasa plugin"),
                         description=_(u""),
                         default=False,)

    picasa_max = schema.Int(title=_(u"Maximum number of photos."),
                            description=_(u"Maximum number of photos to return (maximum value 100)."),
                            default=20,
                            required=True,)

    picasa_desc = schema.Bool(title=_(u"Show Description"),
                              description=_(u"The plugin fetches the title per default. If you also wish to fetch the description, set this option to true."),
                              default=False)


class IHistoryPlugin(Interface):
    """ Enable/Disable History plugin
        http://galleria.io/docs/1.2/plugins/picasa/
    """

    history = schema.Bool(title=_(u"Enable history plugin"),
                          description=_(u""),
                          default=False,)


class IGalleriaSettings(IGeneralSettings, IFlickrPlugin, IPicasaPlugin,
                        IHistoryPlugin, IFaceBookPlugin):
    """The form schema contains all settings."""


class FormGroup1(group.Group):
    label = _(u"Flickr Plugin")
    fields = field.Fields(IFlickrPlugin)


class FormGroup2(group.Group):
    label = _(u"Picasa Plugin")
    fields = field.Fields(IPicasaPlugin)


class FormGroup3(group.Group):
    label = _(u"History Plugin")
    fields = field.Fields(IHistoryPlugin)


class FormGroup4(group.Group):
    label = _(u"FaceBook Plugin")
    fields = field.Fields(IFaceBookPlugin)
